### PR TITLE
Handle Type Mismatch in Array Initialization

### PR DESCRIPTION
**Pull request by HarshitaKalani**
Fixes: https://github.com/lfortran/lfortran/issues/3315

### CODE REVIEW

**Comment by HarshitaKalani** [comment]
This PR fixes the issue for real and integer values. As of now, I'm throwing error for complex arrays as:
```fortran
program main
    complex(8) :: E(3) = [(1, 1), (2, 2), (3, 3)]
    print*, E
end
```
```console
$ lfortran b.f90
semantic error: Type mismatch in array initialization
 --> b.f90:6:5
  |
6 |     complex(8) :: E(3) = [(1, 1), (2, 2), (3, 3)]
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

This can be fixed in another PR.

**Review by certik — APPROVED** [review]
I think this is good.

**Comment by certik** [comment]
Conflicts have to be resolved.

File: src/lfortran/semantics/asr_implicit_cast_rules.h [hunk 1]
```diff
@@ -161,32 +163,34 @@ class ImplicitCastRules {
...
-                value = (ASR::expr_t *)ASR::make_RealConstant_t(al, a_loc,
-                    rval, dest_type2);
+                if (ASR::is_a<ASR::RealConstant_t>(*value)) {
+                    ASR::RealConstant_t *r = ASR::down_cast<ASR::RealConstant_t>(value);
+                    double rval = r->m_r;
+                    value = (ASR::expr_t *)ASR::make_RealConstant_t(al, a_loc,
+                        rval, dest_type2);
+              } else {
+                    value = nullptr;
+              }
```
Comment (HarshitaKalani) [inline src/lfortran/semantics/asr_implicit_cast_rules.h:188]: Here

**Pull request merged**